### PR TITLE
DOP-3187: Create hook for fetching from Hapley API, add global error toast

### PR DESCRIPTION
**Pull request by drew-beckmen**
### Ticket

[DOP-3187](https://jira.mongodb.org/browse/DOP-3187)

### Description

This PR includes a wrapper around the default LG warning toast that sources data from an `ErrorContext`. Using the `setError` from `ErrorContext`, we'll be able to set errors and reuse the same error component across the application. By default, the global error toast disappears after 5 seconds on the screen.

I've also refactored the `fetch` call out of the `SideNav` and `Repo` component into a custom hook. The hook allows us to reuse code in the `fetch` process, namely around error handling. I've leveraged TypeScript generics so that each component that leverages the hook can pass an interface that describes the structure of the API response.

The hook implements the following behavior when handling error responses for GET requests:

- 404 -> redirect to the 404 page
- 401 unauthorized -> redirect to homepage and display error toast with unauthorized message
- all other non-200 responses -> display error toast with message on the user's current page

### Screenshot

Screen Shot 2022-08-11 at 9 59 16 AM

### CODE REVIEW

File: frontend/src/components/error/ApiErrorToast.tsx [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+import { useContext } from 'react';
+import { ErrorContext } from './api-error-context';
+import Toast from '@leafygreen-ui/toast';
+
+const ApiErrorToast: React.FC = () => {
+  const { error, removeError } = useContext(ErrorContext);
+  return (
+    <Toast
+      variant="warning"
```
Comment (mmeigs) [inline frontend/src/components/error/ApiErrorToast.tsx:10]: Just learned this with LG: they have these as constants you can import. So, import { Variant } from toast as well, and then you can use Variant.Warning

File: frontend/src/hooks/use-hapley-api.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  const [data, setData] = useState(null);
+  const [loading, setLoading] = useState(true);
+  const { setError } = useContext(ErrorContext);
+
+  useEffect(() => {
+    const fetchData = async () => {
+      try {
+        const response = await fetch(`${process.env.GATSBY_API_URL}${endpoint}`);
+        if (!response.ok) {
+          throw new Error(response.statusText);
```
Comment (rayangler) [inline frontend/src/hooks/use-hapley-api.ts:19]: I think this is the general text dependent on the status code, and isn't quite specific to any particular issue. Is it possible to use the error message that we're returning from the backend?

File: frontend/src/hooks/use-hapley-api.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import { useEffect, useState, useContext } from 'react';
+import { ErrorContext } from '../components/error/api-error-context';
+
+interface ApiHookProps<T> {
+  data: T | null;
+  loading: boolean;
+}
+
+function useHapleyApi<ApiPayload extends object>(endpoint: string): ApiHookProps<ApiPayload> {
```
Comment (mmeigs) [inline frontend/src/hooks/use-hapley-api.ts:9]: This looks good. For other methods, are we planning on making this a one-stop shop for that later on? ie adding an options object as an argument to handle them all

**Review by rayangler — COMMENTED** [review]
Haven't gotten the chance to play around with this locally yet, but will revisit soon. Mostly just small nits for now.

I'm also curious: is this something that was provided to us by design? I just want to make sure that the design team knows about this since this is a UX change

**Comment by drew-beckmen** [comment]
> I'm also curious: is this something that was provided to us by design? I just want to make sure that the design team knows about this since this is a UX change

It didn't come from design, but I also don't expect this to be a state that users can get to on their own (through a normal flow). This error toast is just to catch unexpected error, such as a state where our API is experiencing issues.

**Comment by rayangler** [comment]
> It didn't come from design, but I also don't expect this to be a state that users can get to on their own (through a normal flow). This error toast is just to catch unexpected error, such as a state where our API is experiencing issues.

I think this is still worth sharing with them to make sure we're on the same page. Especially if this toast is for presenting unexpected/unhandled errors, we should make sure this is the desired UX.

File: frontend/src/components/error/ApiErrorToast.tsx [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+      title="Error processing your request"
+      body={`${error}. Please contact docs platform for assistance.`}
```
[the remark this answers is not in the document]
Comment (drew-beckmen) [reply frontend/src/components/error/ApiErrorToast.tsx:12]: will modify so it's a bit clearer & more formal!

File: frontend/src/hooks/use-hapley-api.ts (hunk 2 above)
Comment (rayangler) [reply frontend/src/hooks/use-hapley-api.ts:19]: I wrote this comment before realizing this was for unexpected errors. Good to ignore if no longer relevant

Comment (drew-beckmen) [reply frontend/src/hooks/use-hapley-api.ts:19]: Cool - I should have clarified that I don't anticipate this hook to be the final pattern for _all_ requests to the API

File: frontend/src/components/SideNav.tsx [hunk 5]
```diff
@@ -1,10 +1,11 @@
+import React from 'react';
 import { SideNav as LGSideNav, SideNavItem } from '@leafygreen-ui/side-nav';
-import { useEffect, useState } from 'react';
+import { ReposResponse } from '../interfaces/apiResponses';
```
[the remark this answers is not in the document]
Comment (drew-beckmen) [reply frontend/src/components/SideNav.tsx:3]: We can make a ticket in the backlog to address this via the linter, but I'll adjust the ordering in the components touched in this PR! thanks for catching!

File: frontend/src/hooks/use-hapley-api.ts (hunk 3 above)
Comment (mmeigs) [reply frontend/src/hooks/use-hapley-api.ts:9]: oh, wow, yeah, delete. :)

File: frontend/src/components/SideNav.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (drew-beckmen) [reply frontend/src/components/SideNav.tsx:3]: Here's the backlog ticket: https://jira.mongodb.org/browse/DOP-3191

**Comment by rayangler** [comment]
Reusable Toast component was added in #47. Error toast messages for unhandled errors were added in #70. Closing this PR

**Pull request closed without merging**